sourced/ml/cmd/repos2bow.py

```diff
@@ -4,50 +4,99 @@
+    start_point = start_point.link(Moder(args.mode)) \
+        .link(PartitionBy.maybe(args.num_iterations, custom_mapping)) \
+        .link(Cacher.maybe("DISK_ONLY"))
+    for num_part in range(args.num_iterations):
+        log.info("Running job %s of %s", num_part + 1, args.num_iterations)
+        selected_part = start_point \
+            .link(PartitionSelector(num_part))  \
+            .link(Repartitioner.maybe(args.partitions, args.shuffle)) \
+            .link(Cacher.maybe(args.persist))
+        if cache_hook is not None:
+            selected_part.link(cache_hook()).execute()
+        uast_extractor = selected_part \
+            .link(UastRow2Document()) \
+            .link(Cacher.maybe(args.persist))
+        log.info("Collecting distinct documents ...")
+        documents = uast_extractor \
+            .link(FieldsSelector([Uast2BagFeatures.Columns.document])) \
+            .link(Distinct()) \
+            .link(Collector()) \
+            .execute()
+        selected_part.unpersist()
+        documents = set(row.document for row in documents)
+        reduced_doc_index = {
+            key: document_index[key] for key in document_index if key in documents}
+        document_indexer = Indexer(Uast2BagFeatures.Columns.document, root.session.sparkContext,
+                                   reduced_doc_index)
+        log.info("Processing %s distinct documents", len(documents))
+        bags = uast_extractor \
+            .link(UastDeserializer()) \
+            .link(Uast2BagFeatures(extractors)) \
+            .link(BagFeatures2TermFreq()) \
+            .link(Cacher.maybe(args.persist))
+        log.info("Extracting UASTs and collecting distinct tokens ...")
+        tokens = bags \
+            .link(FieldsSelector([Uast2BagFeatures.Columns.token])) \
+            .link(Distinct()) \
+            .link(Collector()) \
+            .execute()
+        uast_extractor.unpersist()
+        tokens = set(row.token for row in tokens)
+        reduced_token_freq = {key: df_model[key] for key in df_model.df if key in tokens}
+        reduced_token_index = {key: df_model.order[key] for key in df_model.df if key in tokens}
+        log.info("Processing %s distinct tokens", len(reduced_token_freq))
+        log.info("Indexing by document and token ...")
+        bags_writer = bags \
+            .link(TFIDF(reduced_token_freq, df_model.docs, root.session.sparkContext)) \
+            .link(document_indexer) \
+            .link(Indexer(Uast2BagFeatures.Columns.token, root.session.sparkContext,
+                          reduced_token_index))
+        if save_hook is not None:
+            bags_writer = bags_writer \
+                .link(Repartitioner.maybe(args.partitions, args.shuffle, multiplier=10)) \
+                .link(save_hook())
+        bow = args.bow.split(".asdf")[0] + "_" + str(num_part + 1) + ".asdf"
+        bags_writer \
+            .link(PartitionBy.maybe(
```

Comment:
"PartitionBy" should be renamed to "Partitioner"

Reply:
done